### PR TITLE
Return status

### PR DESCRIPTION
**Pull request by tamarwo16**

### CODE REVIEW

File: client.h [hunk 1]
```diff
@@ -29,7 +31,7 @@ class CheckpointConsumer
 
   // Synchronization and cleanup
   void ReplicationDone(ConsumerState state, const std::string& error);
```
Comment (michaelpanpliops) [inline client.h:33]: I think we should use RepStatus in the callback second parameter.

File: pliops/communication.cc [hunk 2]
```diff
@@ -34,10 +34,11 @@ int Connection<ConnectionType::TCP_SOCKET>::Send(const char* key, uint32_t key_s
                           MSG_NOSIGNAL);
     if (bytes_sent == 0 && errno == 0) {
       // Connection closed by other party (EOF).
-      return 1;
+      logger->Log(Severity::ERROR, FormatString("Connection closed by other party (EOF).\n"));
+      return RepStatus(Code::REPLICATOR_FAILURE, Severity::ERROR, "Connection closed by other party (EOF).\n");
```
Comment (michaelpanpliops) [inline pliops/communication.cc:38]: Code::NETWORK_FAILURE

File: pliops/communication.cc [hunk 3]
```diff
@@ -126,49 +127,49 @@ int Accept(Connection<ConnectionType::TCP_SOCKET>& listen_c,
...
+    return RepStatus(Code::NETWORK_FAILURE, Severity::ERROR, FormatString("Socket accepting failed: %d\n", errno));
   }
 
   // SO_RCVTIMEO
   if (setsockopt(connfd, SOL_SOCKET, SO_RCVTIMEO, &tv_timeout, sizeof(tv_timeout)) < 0) {
-    logger->Log(LogLevel::ERROR, FormatString("Failed connecting socket: setsockopt (set receive timeout) \n"));
+    logger->Log(Severity::ERROR, FormatString("Failed connecting socket: setsockopt (set receive timeout) \n"));
     close(connfd);
-    return -1;
+    return RepStatus(Code::TIMEOUT, Severity::ERROR, FormatString("Failed connecting socket: setsockopt (set receive timeout)\n"));
```
Comment (michaelpanpliops) [inline pliops/communication.cc:148]: Code::NETWORK_FAILURE

File: pliops/communication.cc [hunk 4]
```diff
@@ -126,49 +127,49 @@ int Accept(Connection<ConnectionType::TCP_SOCKET>& listen_c,
...
 }
 
   // SO_SNDTIMEO
   if (setsockopt(connfd, SOL_SOCKET, SO_SNDTIMEO, &tv_timeout, sizeof(tv_timeout)) < 0) {
-    logger->Log(LogLevel::ERROR, FormatString("Failed connecting socket: setsockopt (set send timeout)\n"));
+    logger->Log(Severity::ERROR, FormatString("Failed connecting socket: setsockopt (set send timeout)\n"));
     close(connfd);
-    return -1;
+    return RepStatus(Code::TIMEOUT, Severity::ERROR, FormatString("Failed connecting socket: setsockopt (set send timeout)\n"));
```
Comment (michaelpanpliops) [inline pliops/communication.cc:155]: Code::NETWORK_FAILURE

File: pliops/communication.cc [hunk 5]
```diff
@@ -126,49 +127,49 @@ int Accept(Connection<ConnectionType::TCP_SOCKET>& listen_c,
...
 {
   int listenfd = 0, connfd = 0;
   struct sockaddr_in serv_addr;
 
   listenfd = socket(AF_INET, SOCK_STREAM, 0);
   if (-1 == listenfd) {
-    logger->Log(LogLevel::ERROR, FormatString("Socket creation failed: %d\n", errno));
-    return -1;
+    logger->Log(Severity::ERROR, FormatString("Socket creation failed: %d\n", errno));
+    return RepStatus(Code::REPLICATOR_FAILURE, Severity::ERROR, FormatString("Socket creation failed: %d\n", errno));
```
Comment (michaelpanpliops) [inline pliops/communication.cc:172]: Code::NETWORK_FAILURE

File: pliops/communication.cc [hunk 6]
```diff
@@ -220,16 +221,16 @@ int Connect(const std::string& destination_ip, const uint32_t destination_port,
 
   // SO_RCVTIMEO
   if (setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &tv_timeout, sizeof(tv_timeout)) < 0) {
-      logger->Log(LogLevel::ERROR, FormatString("Failed connecting socket: setsockopt (set receive timeout) \n"));
+      logger->Log(Severity::ERROR, FormatString("Failed connecting socket: setsockopt (set receive timeout) \n"));
       close(sockfd);
-      return -1;
+      return RepStatus(Code::TIMEOUT, Severity::ERROR, FormatString("Failed connecting socket: setsockopt (set receive timeout) \n"));
```
Comment (michaelpanpliops) [inline pliops/communication.cc:226]: Code::NETWORK_FAILURE

File: pliops/consumer.cc [hunk 7]
```diff
@@ -178,12 +183,12 @@ void Consumer::StopImpl()
 
   // Tell threads to exit
   kill_ = true;
-  communication_thread_->join();
+  if (communication_thread_) communication_thread_->join();
```
Comment (michaelpanpliops) [inline pliops/consumer.cc:186]: Where do you set communication_thread_ and other threads to null?

File: pliops/consumer.cc [hunk 8]
```diff
@@ -204,31 +209,31 @@ int Consumer::Stop()
 
   std::future<void>* future = new std::future<void>;
   *future = std::async(std::launch::async, &Consumer::StopImpl, this);
-  if (future->wait_for(10s) == std::future_status::timeout) {
-    logger->Log(LogLevel::ERROR, "Stop failed, some threads are stuck.\n");
-    // If the app will try destroy the Producer object, it will crash
-    return -1;
+  if (future->wait_for(std::chrono::milliseconds(timeout_msec_) + 10s) == std::future_status::timeout) {
```
Comment (michaelpanpliops) [inline pliops/consumer.cc:212]: We will need to change it in my merge

File: pliops/consumer.h [hunk 9]
```diff
@@ -26,11 +28,11 @@ class Consumer {
...
   explicit Consumer(uint64_t timeout_msec, IKvPairSerializer& kv_pair_serializer);
   virtual ~Consumer();
-  int Start(const std::string& replica_path, uint16_t& port,
+  RepStatus Start(const std::string& replica_path, uint16_t& port,
             std::function<void(ConsumerState, const std::string&)>& done_callback);
-  int Stop();
-  int GetState(ConsumerState& state, std::string& error);
-  int GetStats(uint64_t& num_kv_pairs, uint64_t& num_bytes);
+  RepStatus Stop();
+  RepStatus GetState(ConsumerState& state, std::string& error);
```
Comment (michaelpanpliops) [inline pliops/consumer.h:34]: I think GetState and SetState should also use RepStatus

File: pliops/producer.h [hunk 10]
```diff
@@ -28,13 +31,13 @@ class Producer {
...
-  int Start(const std::string& ip, uint16_t port,
+  RepStatus OpenShard(const std::string& shard_path);
+  RepStatus Start(const std::string& ip, uint16_t port,
             uint32_t max_num_of_threads, uint32_t parallelism, uint64_t timeout_msec,
             std::function<void(ProducerState, const std::string&)>& done_callback);
-  int Stop();
-  int GetState(ProducerState& state, std::string& error);
-  int GetStats(uint64_t& num_kv_pairs, uint64_t& num_bytes);
+  RepStatus Stop();
+  RepStatus GetState(ProducerState& state, std::string& error);
```
Comment (michaelpanpliops) [inline pliops/producer.h:39]: Also here, RepStatus => the second parameter in SetState and GetState

File: pliops/producer.cc [hunk 11]
```diff
@@ -299,30 +299,30 @@ int Producer::Stop()
   std::future<void>* future = new std::future<void>;
   *future = std::async(std::launch::async, &Producer::StopImpl, this);
   if (future->wait_for(10s) == std::future_status::timeout) {
```
Comment (michaelpanpliops) [inline pliops/producer.cc:301]: Here we need to change timeout as well

File: pliops/producer.cc [hunk 12]
```diff
@@ -277,7 +277,7 @@ void Producer::StopImpl()
   // We need to signal communication thread with poison pill
   // because it could be waiting on queue
   message_queue_->enqueue({"", ""});
-  communication_thread_->join();
+  if (communication_thread_) communication_thread_->join();
```
Comment (michaelpanpliops) [inline pliops/producer.cc:280]: Where do you set it to null? Also, should you check message_queue_ as well?

File: pliops/status.h [hunk 13]
```diff
@@ -0,0 +1,31 @@
+#ifndef STATUS_H
```
Comment (michaelpanpliops) [inline pliops/status.h:1]: #pragma once

File: pliops/status.h [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+#include <string>
+
+#include "defs.h"
+
+namespace Replicator {
+
+enum class Code : int {
+    OK = 0,
+    TIMEOUT = 1,
```
Comment (michaelpanpliops) [inline pliops/status.h:12]: Maybe TIMEOUT_FAILURE ?

File: pliops/status.h [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    NETWORK_FAILURE = 2,
+    DB_FAILURE = 3,
+    REPLICATOR_FAILURE
```
Comment (michaelpanpliops) [inline pliops/status.h:15]: = 4

File: pliops/status.cc [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+#include "status.h"
+
+namespace Replicator {
+
+Status::Status(Code code, Severity severity, const std::string& message)
+    : severity_(severity), message_(message), code_(code) {}
+
+std::string Status::ToString() const {
+    std::string severity_str;
+    switch (severity_) {
```
Comment (michaelpanpliops) [inline pliops/status.cc:11]: 1. FATAL and OK are missing.
2. I would put this translation into defs.h

File: server.h [hunk 17]
```diff
@@ -24,14 +24,14 @@ class CheckpointProducer
...
   // Client requests processing methods
-  int CreateCheckpoint(const CreateCheckpointRequest& req, CreateCheckpointResponse& res);
-  int StartStreaming(const StartStreamingRequest& req, StartStreamingResponse& res, uint64_t timeout_msec);
-  int GetStatus(const GetStatusRequest& req, GetStatusResponse& res);
+  RepStatus CreateCheckpoint(const CreateCheckpointRequest& req, CreateCheckpointResponse& res);
+  RepStatus StartStreaming(const StartStreamingRequest& req, StartStreamingResponse& res, uint64_t timeout_msec);
+  RepStatus GetStatus(const GetStatusRequest& req, GetStatusResponse& res);
 
   // Synchronization and cleanup
   void ReplicationDone(ProducerState state, const std::string& error);
```
Comment (michaelpanpliops) [inline server.h:32]: Use RepStatus in second parameter

File: server.cc [hunk 18]
```diff
@@ -105,137 +105,142 @@ int CheckpointProducer::StartStreaming(
...
+  std::function<RepStatus(const GetStatusRequest&, GetStatusResponse&)>
     get_status_cb = std::bind(&CheckpointProducer::GetStatus, &cp, _1, _2); 
   while(!cp.IsClientDone()) {
     rc = rpc.ProcessCommand(get_status_cb);
-    if (rc) {
-      logger->Log(LogLevel::ERROR, FormatString("CheckpointProducer::GetStatus failed\n"));
-      return -1;
+    if (!rc.IsOk()) {
+      logger->Log(Severity::ERROR, FormatString("CheckpointProducer::GetStatus failed\n"));
+      cp.DestroyCheckpoint();
```
Comment (michaelpanpliops) [inline server.cc:228]: Can we do it with the cp destructor?

File: server.cc [hunk 19]
```diff
@@ -105,137 +105,142 @@ int CheckpointProducer::StartStreaming(
...
+RepStatus CheckpointProducer::GetStatus(
                           const GetStatusRequest& req,
                           GetStatusResponse& res)
 {
   // We expect to get the same checkpoint_id as we provided in the CreateCheckpoint call
   if (req.checkpoint_id != checkpoint_id_) {
-    logger->Log(LogLevel::ERROR, FormatString("Invalid checkpoint id\n"));
-    return -1;
+    logger->Log(Severity::ERROR, FormatString("Invalid checkpoint id\n"));
+    DestroyCheckpoint();
```
Comment (michaelpanpliops) [inline server.cc:129]: Why do we need to call DestroyCheckpoint here? I think it should be cleaned by the caller.

Comment (tamarwo16) [reply server.cc:129]: You are right, it's redundant here. Removed.

File: client.cc [hunk 20]
```diff
@@ -154,8 +157,7 @@ RepStatus CheckReplicationStatus(RpcChannel& rpc, bool& done)
 
   // Get the state from the consumer
   ConsumerState consumer_state;
-  std::string error;
-  auto rc = consumer_->ConsumerImpl().GetState(consumer_state, error);
+  auto rc = consumer_->ConsumerImpl().GetState(consumer_state);
```
Comment (michaelpanpliops) [inline client.cc:160]: Also here, I thought of returning RepStatus  as a second parameter

File: pliops/consumer.cc [hunk 21]
```diff
@@ -221,12 +221,11 @@ RepStatus Consumer::Stop()
...
 }
 
-RepStatus Consumer::GetState(ConsumerState& state, std::string& error)
+RepStatus Consumer::GetState(ConsumerState& state)
 {
   std::lock_guard<std::mutex> lock(state_mutex_);
   state = state_;
-  error = error_;
-  return RepStatus();
+  return rc_;
```
Comment (michaelpanpliops) [inline pliops/consumer.cc:228]: I see what you did. But I think the return value meaning is different. It refers to the function call status, but the rc_ is only refers to the background errors.

File: server.h [hunk 22]
```diff
@@ -21,15 +21,15 @@ class CheckpointProducer
...
-  ~CheckpointProducer() {}
+  ~CheckpointProducer() { DestroyCheckpoint(); }
 
   // Client requests processing methods
   RepStatus CreateCheckpoint(const CreateCheckpointRequest& req, CreateCheckpointResponse& res);
   RepStatus StartStreaming(const StartStreamingRequest& req, StartStreamingResponse& res, uint64_t timeout_msec);
   RepStatus GetStatus(const GetStatusRequest& req, GetStatusResponse& res);
 
   // Synchronization and cleanup
-  void ReplicationDone(ProducerState state, const std::string& error);
```
Comment (michaelpanpliops) [inline server.h:32]: Also here I thought of passing RepStatus in parameter.

File: pliops/producer.cc [hunk 23]
```diff
@@ -309,12 +301,11 @@ RepStatus Producer::Stop()
...
 }
 
-RepStatus Producer::GetState(ProducerState& state, std::string& error)
+RepStatus Producer::GetState(ProducerState& state)
 {
   std::lock_guard<std::mutex> lock(state_mutex_);
   state = state_;
-  error = error_;
-  return RepStatus();
+  return rc_;
```
Comment (michaelpanpliops) [inline pliops/producer.cc:308]: Same comment as in Consumer::GetState

File: pliops/producer.h [hunk 24]
```diff
@@ -34,14 +34,14 @@ class Producer {
...
+            std::function<void(ProducerState)>& done_callback);
   RepStatus Stop();
-  RepStatus GetState(ProducerState& state, std::string& error);
+  RepStatus GetState(ProducerState& state);
   RepStatus GetStats(uint64_t& num_kv_pairs, uint64_t& num_bytes);
 
 private:
   // Callback to be called on completion/error
-  std::function<void(ProducerState, const std::string&)> done_callback_;
+  std::function<void(ProducerState)> done_callback_;
```
Comment (michaelpanpliops) [inline pliops/producer.h:44]: Same comment as in Consumer

File: pliops/status.h [hunk 25]
```diff
@@ -9,12 +10,32 @@ namespace Replicator {
 
 enum class Code : int {
     OK = 0,
```
Comment (michaelpanpliops) [inline pliops/status.h:12]: Why are you using 4space tab? The rest of the project is using 2space tabs.

File: server.cc [hunk 26]
```diff
@@ -185,14 +183,17 @@ RepStatus CheckpointProducer::DestroyCheckpoint() {
   return RepStatus();
 }
 
-void CheckpointProducer::ReplicationDone(ProducerState state, const std::string& error)
+void CheckpointProducer::ReplicationDone(ProducerState state)
```
Comment (michaelpanpliops) [inline server.cc:186]: Also here - second param RepStatus

File: server.cc [hunk 27]
```diff
@@ -126,7 +126,6 @@ RepStatus CheckpointProducer::GetStatus(
   // We expect to get the same checkpoint_id as we provided in the CreateCheckpoint call
   if (req.checkpoint_id != checkpoint_id_) {
     logger->Log(Severity::ERROR, FormatString("Invalid checkpoint id\n"));
-    DestroyCheckpoint();
     return RepStatus(Code::DB_FAILURE, Severity::ERROR, FormatString("Invalid checkpoint id\n"));
```
Comment (michaelpanpliops) [inline server.cc:129]: What about other DestroyCheckpoint in this function? Can you remove them?

Comment (tamarwo16) [reply server.cc:129]: yes

File: client.cc [hunk 28]
```diff
@@ -157,7 +158,8 @@ RepStatus CheckReplicationStatus(RpcChannel& rpc, bool& done)
 
   // Get the state from the consumer
   ConsumerState consumer_state;
-  auto rc = consumer_->ConsumerImpl().GetState(consumer_state);
+  RepStatus consumer_status;
+  auto rc = consumer_->ConsumerImpl().GetState(consumer_state, consumer_status);
```
Comment (michaelpanpliops) [inline client.cc:162]: print consumer_status if in final state

File: pliops/producer.cc [hunk 29]
```diff
@@ -264,9 +264,9 @@ void Producer::StopImpl()
 
   // We need to signal communication thread with poison pill
   // because it could be waiting on queue
-  if (message_queue_ && !EnqueueTimed(message_queue_, {"",""}, kill_, timeout_msec_) ) {
+  if (message_queue_ && !EnqueueTimed(message_queue_, {"",""}, false, timeout_msec_) ) {
```
Comment (michaelpanpliops) [inline pliops/producer.cc:267]: Replace with temp var

**Pull request merged**